File: 2022/swift/02.swift
import Foundation

let filename = "./input/02.in"

enum Throw: Int {
    case rock = 0; case paper = 1; case scissors = 2
}

enum Result: Int {
    case win = 1; case lose = -1; case draw = 0
}

let board: [[Result]] = [[.draw, .lose, .win], [.win, .draw, .lose], [.lose, .win, .draw]]

func normalHand(_ input: String) -> Throw? {
    if input == "A" || input == "X" {
        return .rock
    }
    if input == "B" || input == "Y" {
        return .paper
    }
    if input == "C" || input == "Z" {
        return .scissors
    }

    return nil
}

func handScore(_ hand: Throw) -> Int {
    switch hand {
    case .rock:
        return 1
    case .paper:
        return 2
    case .scissors:
        return 3
    }
}

func normalResult(_ input: String) -> Result? {
    if input == "X" {
        return .lose
    }

    if input == "Y" {
        return .draw
    }

    if input == "Z" {
        return .win
    }

    return nil
}

func main() {
    do {
        let games = try String(contentsOfFile: filename).split(separator: "\n").map { $0.split(separator: " ").map { String($0) }}

        print("Part one:", games.map {
            game in
            game.map { normalHand($0)! }
        }.reduce(0) {
            prev, curr in
            let hand = curr[1]
            let other = curr[0]
            let result = board[hand.rawValue][other.rawValue]

            let score = result == .win ? 6 : result == .draw ? 3 : 0

            return prev + score + handScore(hand)

        })

        print("Part two:", games.map {
            game in
            (normalHand(game[0])!, normalResult(game[1])!)
        }.reduce(0) {
            prev, curr in
            let (other, wanted) = curr

            let opponentResult = wanted.rawValue * -1

            let index = board[other.rawValue].firstIndex { $0.rawValue == opponentResult }!

            let hand = Throw(rawValue: index)!

            let score = wanted == .win ? 6 : wanted == .draw ? 3 : 0

            return prev + score + handScore(hand)

        })

    } catch {
        print(error)
    }
}

main()

File: 2022/src/bin/02.rs
#[derive(Debug, Copy, Clone)]
enum Throw {
    Rock = 0,
    Paper = 1,
    Scissors = 2,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum Result {
    Win = 1,
    Lose = -1,
    Draw = 0,
}
// Ordered as Rock Paper Scissors
const RESULTS: [[Result; 3]; 3] = [
    [Result::Draw, Result::Lose, Result::Win],
    [Result::Win, Result::Draw, Result::Lose],
    [Result::Lose, Result::Win, Result::Draw],
];

impl Result {
    fn score(self: &Result) -> usize {
        match self {
            Result::Win => 6,
            Result::Draw => 3,
            Result::Lose => 0,
        }
    }
    fn from(str: &str) -> Result {
        match str {
            "X" => Result::Lose,
            "Y" => Result::Draw,
            "Z" => Result::Win,
            _ => panic!("Invalid input"),
        }
    }

    fn reverse(self: &Result) -> Result {
        match self {
            Result::Win => Result::Lose,
            Result::Lose => Result::Win,
            Result::Draw => Result::Draw,
        }
    }
}

impl Throw {
    fn score(self: &Throw) -> usize {
        match self {
            Throw::Rock => 1,
            Throw::Paper => 2,
            Throw::Scissors => 3,
        }
    }

    fn from(str: &str) -> Throw {
        match str {
            "A" | "X" => Throw::Rock,
            "B" | "Y" => Throw::Paper,
            "C" | "Z" => Throw::Scissors,
            _ => panic!("Invalid input"),
        }
    }

    fn result(self: &Throw, other: &Throw) -> &Result {
        return &RESULTS[*self as usize][*other as usize];
    }

    fn from_result(self: &Throw, wanted: &Result) -> Throw {
        match &RESULTS[*self as usize].iter().position(|res| res == wanted) {
            Some(0) => Throw::Rock,
            Some(1) => Throw::Paper,
            Some(2) => Throw::Scissors,
            _ => panic!("Invalid position"),
        }
    }
}

#[aoc2022::main(02)]
fn main(input: &str) -> (usize, usize) {
    let games = input
        .split("\n")
        .map(|game| game.split(" ").map(String::from).collect())
        .collect::<Vec<Vec<String>>>();

    let part_one = games.iter().fold(0usize, |prev, curr| {
        let hand = Throw::from(&curr[1]);
        let other = Throw::from(&curr[0]);

        let result = hand.result(&other);
        return prev + hand.score() + result.score();
    });

    let part_two = games.iter().fold(0usize, |prev, curr| {
        let other = Throw::from(&curr[0]);

        let wanted = Result::from(&curr[1]);

        let hand = other.from_result(&wanted.reverse());

        return prev + hand.score() + wanted.score();
    });

    (part_one, part_two)
}
